--- a/net/minecraft/client/renderer/ItemRenderer.java
+++ b/net/minecraft/client/renderer/ItemRenderer.java
@@ -372,6 +372,95 @@
 
     public void renderItemInFirstPerson(AbstractClientPlayer p_187457_1_, float p_187457_2_, float p_187457_3_, EnumHand p_187457_4_, float p_187457_5_, ItemStack p_187457_6_, float p_187457_7_)
     {
+        boolean flag = p_187457_4_ == EnumHand.MAIN_HAND;
+        EnumHandSide enumhandside = flag ? p_187457_1_.getPrimaryHand() : p_187457_1_.getPrimaryHand().opposite();
+        GlStateManager.pushMatrix();
+
+        if (p_187457_6_.isEmpty())
+        {
+            if (flag && !p_187457_1_.isInvisible())
+            {
+                this.renderArmFirstPerson(p_187457_7_, p_187457_5_, enumhandside);
+            }
+        }
+        else if (p_187457_6_.getItem() instanceof ItemMap)
+        {
+            if (flag && this.itemStackOffHand.isEmpty())
+            {
+                this.renderMapFirstPerson(p_187457_3_, p_187457_7_, p_187457_5_);
+            }
+            else
+            {
+                this.renderMapFirstPersonSide(p_187457_7_, enumhandside, p_187457_5_, p_187457_6_);
+            }
+        }
+        else
+        {
+            boolean flag1 = enumhandside == EnumHandSide.RIGHT;
+
+            if (p_187457_1_.isHandActive() && p_187457_1_.getItemInUseCount() > 0 && p_187457_1_.getActiveHand() == p_187457_4_)
+            {
+                int j = flag1 ? 1 : -1;
+
+                switch (p_187457_6_.getItemUseAction())
+                {
+                    case NONE:
+                        this.transformSideFirstPerson(enumhandside, p_187457_7_);
+                        break;
+
+                    case EAT:
+                    case DRINK:
+                        this.transformEatFirstPerson(p_187457_2_, enumhandside, p_187457_6_);
+                        this.transformSideFirstPerson(enumhandside, p_187457_7_);
+                        break;
+
+                    case BLOCK:
+                        this.transformSideFirstPerson(enumhandside, p_187457_7_);
+                        break;
+
+                    case BOW:
+                        this.transformSideFirstPerson(enumhandside, p_187457_7_);
+                        GlStateManager.translate((float)j * -0.2785682F, 0.18344387F, 0.15731531F);
+                        GlStateManager.rotate(-13.935F, 1.0F, 0.0F, 0.0F);
+                        GlStateManager.rotate((float)j * 35.3F, 0.0F, 1.0F, 0.0F);
+                        GlStateManager.rotate((float)j * -9.785F, 0.0F, 0.0F, 1.0F);
+                        float f5 = (float)p_187457_6_.getMaxItemUseDuration() - ((float)this.mc.player.getItemInUseCount() - p_187457_2_ + 1.0F);
+                        float f6 = f5 / 20.0F;
+                        f6 = (f6 * f6 + f6 * 2.0F) / 3.0F;
+
+                        if (f6 > 1.0F)
+                        {
+                            f6 = 1.0F;
+                        }
+
+                        if (f6 > 0.1F)
+                        {
+                            float f7 = MathHelper.sin((f5 - 0.1F) * 1.3F);
+                            float f3 = f6 - 0.1F;
+                            float f4 = f7 * f3;
+                            GlStateManager.translate(f4 * 0.0F, f4 * 0.004F, f4 * 0.0F);
+                        }
+
+                        GlStateManager.translate(f6 * 0.0F, f6 * 0.0F, f6 * 0.04F);
+                        GlStateManager.scale(1.0F, 1.0F, 1.0F + f6 * 0.2F);
+                        GlStateManager.rotate((float)j * 45.0F, 0.0F, -1.0F, 0.0F);
+                }
+            }
+            else
+            {
+                float f = -0.4F * MathHelper.sin(MathHelper.sqrt(p_187457_5_) * (float)Math.PI);
+                float f1 = 0.2F * MathHelper.sin(MathHelper.sqrt(p_187457_5_) * ((float)Math.PI * 2F));
+                float f2 = -0.2F * MathHelper.sin(p_187457_5_ * (float)Math.PI);
+                int i = flag1 ? 1 : -1;
+                GlStateManager.translate((float)i * f, f1, f2);
+                this.transformSideFirstPerson(enumhandside, p_187457_7_);
+                this.transformFirstPerson(enumhandside, p_187457_5_);
+            }
+
+            this.renderItemSide(p_187457_1_, p_187457_6_, flag1 ? ItemCameraTransforms.TransformType.FIRST_PERSON_RIGHT_HAND : ItemCameraTransforms.TransformType.FIRST_PERSON_LEFT_HAND, !flag1);
+        }
+
+        GlStateManager.popMatrix();
     }
 
     /**
