--- a/net/minecraft/client/renderer/RenderGlobal.java
+++ b/net/minecraft/client/renderer/RenderGlobal.java
@@ -743,7 +743,7 @@
             List<Entity> list2 = Lists.<Entity>newArrayList();
             BlockPos.PooledMutableBlockPos blockpos$pooledmutableblockpos = BlockPos.PooledMutableBlockPos.retain();
 
-            for (RenderGlobal.ContainerLocalRenderInformation renderglobal$containerlocalrenderinformation : this.renderInfosEntities)
+            for (RenderGlobal.ContainerLocalRenderInformation renderglobal$containerlocalrenderinformation :  (Iterable<RenderGlobal.ContainerLocalRenderInformation>)this.renderInfosEntities)
             {
                 Chunk chunk = renderglobal$containerlocalrenderinformation.renderChunk.getChunk();
                 ClassInheritanceMultiMap<Entity> classinheritancemultimap = chunk.getEntityLists()[renderglobal$containerlocalrenderinformation.renderChunk.getPosition().getY() / 16];
@@ -908,7 +908,7 @@
 
             label251:
 
-            for (RenderGlobal.ContainerLocalRenderInformation renderglobal$containerlocalrenderinformation1 : this.renderInfosTileEntities)
+            for (RenderGlobal.ContainerLocalRenderInformation renderglobal$containerlocalrenderinformation1 : (Iterable<RenderGlobal.ContainerLocalRenderInformation>)this.renderInfosTileEntities)
             {
                 List<TileEntity> list3 = renderglobal$containerlocalrenderinformation1.renderChunk.getCompiledChunk().getTileEntities();
 
@@ -1560,10 +1560,7 @@
                 GlStateManager.disableFog();
             }
 
-            this.mc.mcProfiler.func_194339_b(() ->
-            {
-                return "render_" + p_lambda$renderBlockLayer$0_0_;
-            });
+            this.mc.mcProfiler.startSection("render_" + blockLayerIn);
             this.renderBlockLayer(blockLayerIn);
             this.mc.mcProfiler.endSection();
             return l;
