--- a/net/minecraft/entity/projectile/EntityArrow.java
+++ b/net/minecraft/entity/projectile/EntityArrow.java
@@ -7,6 +7,7 @@
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.Minecraft;
 import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
@@ -92,6 +93,14 @@
 
         if (shooter instanceof EntityPlayer)
         {
+        	if(shooter.getEntityId() == Minecraft.getMinecraft().player.getEntityId()){
+        		//VIVE SET SPAWNPOS
+        		Vec3d aimSource = Minecraft.getMinecraft().roomScale.getControllerPos_World(1);
+        		Vec3d controllerDirection = Minecraft.getMinecraft().bowTracker.getAimVector();
+        		this.setPosition(aimSource.x + controllerDirection.x * 0.1f,aimSource.y + controllerDirection.y * 0.1f,aimSource.z + controllerDirection.z * 0.1f);
+        		///
+        	}
+            
             this.pickupStatus = EntityArrow.PickupStatus.ALLOWED;
         }
     }
