--- a/net/minecraft/entity/projectile/EntityFireball.java
+++ b/net/minecraft/entity/projectile/EntityFireball.java
@@ -1,7 +1,12 @@
 package net.minecraft.entity.projectile;
 
+import com.mtbs3d.minecrift.api.NetworkHelper;
+import com.mtbs3d.minecrift.api.VivePlayer;
+
+import net.minecraft.client.Minecraft;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.util.DamageSource;
@@ -65,7 +70,20 @@
         super(worldIn);
         this.shootingEntity = shooter;
         this.setSize(1.0F, 1.0F);
-        this.setLocationAndAngles(shooter.posX, shooter.posY, shooter.posZ, shooter.rotationYaw, shooter.rotationPitch);
+        
+        //VIVE USE CONTROLLER
+    	VivePlayer vp = NetworkHelper.vivePlayers.get(shooter.getUniqueID());
+    	if (vp !=null && vp.isVR())
+    	{
+    		Vec3d aim = vp.getControllerDir(0);
+    		Vec3d aimSource = vp.getControllerPos(0);
+    		float pitch = (float)Math.toDegrees(Math.asin(aim.y/aim.lengthVector()));
+    		float yaw = (float)Math.toDegrees(Math.atan2(aim.x, -aim.z));   
+    		this.setLocationAndAngles(aimSource.x + aim.x*0.6f, aimSource.y+aim.y*0.6f, aimSource.z+aim.z*0.6f, yaw, pitch);
+        } else {
+        	this.setLocationAndAngles(shooter.posX, shooter.posY, shooter.posZ, shooter.rotationYaw, shooter.rotationPitch);
+        }
+        
         this.setPosition(this.posX, this.posY, this.posZ);
         this.motionX = 0.0D;
         this.motionY = 0.0D;
