--- a/net/minecraft/entity/projectile/EntityFireball.java
+++ b/net/minecraft/entity/projectile/EntityFireball.java
@@ -1,7 +1,9 @@
 package net.minecraft.entity.projectile;
 
+import net.minecraft.client.Minecraft;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.util.DamageSource;
@@ -65,7 +67,18 @@
         super(worldIn);
         this.shootingEntity = shooter;
         this.setSize(1.0F, 1.0F);
-        this.setLocationAndAngles(shooter.posX, shooter.posY, shooter.posZ, shooter.rotationYaw, shooter.rotationPitch);
+        
+        //VIVE USE CONTROLLER
+        if(shooter.getEntityId() == Minecraft.getMinecraft().player.getEntityId()){
+        	float yaw = Minecraft.getMinecraft().roomScale.getControllerYaw_World(0);
+        	float pitch = -Minecraft.getMinecraft().roomScale.getControllerPitch_World(0);
+        	Vec3d aimSource = Minecraft.getMinecraft().roomScale.getControllerPos_World(0);
+        	this.setLocationAndAngles(aimSource.x	,aimSource.y, aimSource.z, yaw, pitch);
+        	//END VIVE
+        } else {
+        	this.setLocationAndAngles(shooter.posX, shooter.posY, shooter.posZ, shooter.rotationYaw, shooter.rotationPitch);
+        }
+        
         this.setPosition(this.posX, this.posY, this.posZ);
         this.motionX = 0.0D;
         this.motionY = 0.0D;
