--- a/net/optifine/entity/model/anim/ExpressionParser.java
+++ b/net/optifine/entity/model/anim/ExpressionParser.java
@@ -242,7 +242,7 @@
         }
     }
 
-    private FunctionType getFunctionType(Token token, Deque<Token> deque) throws ParseException
+    private FunctionType getFunctionType(Token tokenin, Deque<Token> deque) throws ParseException
     {
         Token token = deque.peek();
 
@@ -254,7 +254,7 @@
         }
         else
         {
-            FunctionType functiontype = FunctionType.parse(token.getText());
+            FunctionType functiontype = FunctionType.parse(tokenin.getText());
 
             if (functiontype == null)
             {
@@ -271,7 +271,7 @@
         }
     }
 
-    private IExpression makeFunction(FunctionType type, Deque<Token> deque) throws ParseException
+    private IExpression makeFunction(FunctionType type, Deque<Token> dequein) throws ParseException
     {
         if (type.getParameterCount(new IExpression[0]) == 0)
         {
@@ -279,20 +279,20 @@
         }
         else
         {
-            Token token = deque.poll();
-            Deque<Token> deque = getGroup(deque, TokenType.BRACKET_CLOSE, true);
+            Token token = dequein.poll();
+            Deque<Token> deque = getGroup(dequein, TokenType.BRACKET_CLOSE, true);
             IExpression[] aiexpression = this.parseExpressions(deque);
             return makeFunction(type, aiexpression);
         }
     }
 
-    private IExpression[] parseExpressions(Deque<Token> deque) throws ParseException
+    private IExpression[] parseExpressions(Deque<Token> dequein) throws ParseException
     {
         List<IExpression> list = new ArrayList<IExpression>();
 
         while (true)
         {
-            Deque<Token> deque = getGroup(deque, TokenType.COMMA, false);
+            Deque<Token> deque = getGroup(dequein, TokenType.COMMA, false);
             IExpression iexpression = this.parseInfix(deque);
 
             if (iexpression == null)
@@ -363,13 +363,13 @@
         }
     }
 
-    private IExpression makeBracketed(Token token, Deque<Token> deque) throws ParseException
+    private IExpression makeBracketed(Token token, Deque<Token> dequein) throws ParseException
     {
-        Deque<Token> deque = getGroup(deque, TokenType.BRACKET_CLOSE, true);
+        Deque<Token> deque = getGroup(dequein, TokenType.BRACKET_CLOSE, true);
         return this.parseInfix(deque);
     }
 
-    private static Deque<Token> getGroup(Deque<Token> deque, TokenType tokenTypeEnd, boolean tokenEndRequired) throws ParseException
+    private static Deque<Token> getGroup(Deque<Token> dequein, TokenType tokenTypeEnd, boolean tokenEndRequired) throws ParseException
     {
         Deque<Token> deque = new ArrayDeque<Token>();
         int i = 0;
