--- a/net/minecraft/init/Bootstrap.java
+++ b/net/minecraft/init/Bootstrap.java
@@ -47,12 +47,15 @@
 import net.minecraft.item.ItemMonsterPlacer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.item.crafting.ShapedRecipes;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTUtil;
 import net.minecraft.potion.Potion;
 import net.minecraft.potion.PotionHelper;
 import net.minecraft.potion.PotionType;
 import net.minecraft.server.DebugLoggingPrintStream;
+import net.optifine.reflect.Reflector;
 import net.minecraft.stats.StatList;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.tileentity.TileEntityDispenser;
@@ -60,6 +63,8 @@
 import net.minecraft.tileentity.TileEntitySkull;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.LoggingPrintStream;
+import net.minecraft.util.NonNullList;
+import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.SoundEvent;
 import net.minecraft.util.StringUtils;
@@ -267,6 +272,11 @@
         };
         BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.LAVA_BUCKET, ibehaviordispenseitem);
         BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.WATER_BUCKET, ibehaviordispenseitem);
+        if (Reflector.forgeExists()) {
+        	BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.MILK_BUCKET, (IBehaviorDispenseItem)Reflector.call(Reflector.DispenseFluidContainer_getInstance));
+        	BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.BUCKET, (IBehaviorDispenseItem)Reflector.call(Reflector.DispenseFluidContainer_getInstance));
+        }
+        else
         BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.BUCKET, new BehaviorDefaultDispenseItem()
         {
             private final BehaviorDefaultDispenseItem dispenseBehavior = new BehaviorDefaultDispenseItem();
@@ -495,13 +505,15 @@
         if (!alreadyRegistered)
         {
             alreadyRegistered = true;
-            redirectOutputToLog();
+			if (!forgeExistsSafe()) // We're too early in init to load Reflector, so we gotta do this differently
+				redirectOutputToLog();
             SoundEvent.registerSounds();
             Block.registerBlocks();
             BlockFire.init();
             Potion.registerPotions();
             Enchantment.registerEnchantments();
             Item.registerItems();
+            
             PotionType.registerPotionTypes();
             PotionHelper.init();
             EntityList.init();
@@ -514,6 +526,32 @@
                 LOGGER.error("Errors with built-in recipes!");
             }
 
+            //VIVECRAFT
+            ItemStack is = new ItemStack(Items.LEATHER_BOOTS);
+            is.setStackDisplayName("Jump Boots");
+            is.getTagCompound().setBoolean("Unbreakable", true);
+            is.getTagCompound().setInteger("HideFlags",4);
+
+            ItemStack is2 = new ItemStack(Items.SHEARS);
+            is2.setStackDisplayName("Climb Claws");
+            is2.getTagCompound().setBoolean("Unbreakable", true);
+            is2.getTagCompound().setInteger("HideFlags",4);
+            
+            ShapedRecipes boot = new ShapedRecipes("", 1, 2, NonNullList.from(Ingredient.EMPTY,Ingredient.fromItem(Items.LEATHER_BOOTS), Ingredient.fromStacks(new ItemStack(Blocks.SLIME_BLOCK))), is);
+            ShapedRecipes claw = new ShapedRecipes("", 3, 2, NonNullList.from(Ingredient.EMPTY,Ingredient.fromItem(Items.SPIDER_EYE),Ingredient.EMPTY,Ingredient.fromItem(Items.SPIDER_EYE),Ingredient.fromItem(Items.SHEARS),Ingredient.EMPTY,Ingredient.fromItem(Items.SHEARS)), is2);
+
+            //
+            if(Reflector.forgeExists()){
+            	Reflector.call(Reflector.GameRegistry_addShapedRecipe, new ResourceLocation("vivecraftforgeextensions:jump_boots"), (ResourceLocation) null, is, new Object[]{  "B", "S", 'S', Blocks.SLIME_BLOCK, 'B', Items.LEATHER_BOOTS});
+            	Reflector.call(Reflector.GameRegistry_addShapedRecipe, new ResourceLocation("vivecraftforgeextensions:climb_claws"), (ResourceLocation) null, is2, new Object[]{  "S S", "C C", 'S', Items.SPIDER_EYE, 'C', Items.SHEARS});
+            }
+            else
+            {
+            	CraftingManager.register("jumpboots", boot);
+            	CraftingManager.register("climbclaws", claw);
+            }
+            //
+
             StatList.init();
 
             if (LOGGER.isDebugEnabled())
@@ -530,6 +568,8 @@
                     LOGGER.error("Errors with built-in loot tables");
                 }
             }
+
+			Reflector.callVoid(Reflector.GameData_vanillaSnapshot);
         }
     }
 
@@ -550,6 +590,16 @@
         }
     }
 
+    public static boolean forgeExistsSafe() {
+		try {
+			// This should be something safe to load really early during init, ideally having no imports at all
+			Class.forName("net.minecraftforge.fml.common.versioning.InvalidVersionSpecificationException");
+		} catch (ClassNotFoundException e) {
+			return false;
+		}
+		return true;
+	}
+
     public static void printToSYSOUT(String message)
     {
         SYSOUT.println(message);
