--- a/net/minecraft/init/Items.java
+++ b/net/minecraft/init/Items.java
@@ -220,6 +220,8 @@
     public static final Item BEETROOT;
     public static final Item BEETROOT_SOUP;
     public static final Item TOTEM_OF_UNDYING;
+    public static final Item EATME_CAKE;
+    public static final Item DRIINKME_BOTTLE;
     public static final Item SHULKER_SHELL;
     public static final Item IRON_NUGGET;
     public static final Item KNOWLEDGE_BOOK;
@@ -453,6 +455,8 @@
             BEETROOT = getRegisteredItem("beetroot");
             BEETROOT_SOUP = getRegisteredItem("beetroot_soup");
             TOTEM_OF_UNDYING = getRegisteredItem("totem_of_undying");
+            EATME_CAKE = getRegisteredItem("eatme_cake");
+            DRIINKME_BOTTLE = getRegisteredItem("drinkme_bottle");
             SHULKER_SHELL = getRegisteredItem("shulker_shell");
             IRON_NUGGET = getRegisteredItem("iron_nugget");
             KNOWLEDGE_BOOK = getRegisteredItem("knowledge_book");
