--- a/net/minecraft/client/renderer/entity/RenderManager.java
+++ b/net/minecraft/client/renderer/entity/RenderManager.java
@@ -261,7 +261,7 @@
 
         if (render == null && entityClass != Entity.class)
         {
-            render = this.<T>getEntityClassRenderObject(entityClass.getSuperclass());
+            render = this.getEntityClassRenderObject((Class<? extends Entity>) entityClass.getSuperclass());
             this.entityRenderMap.put(entityClass, render);
         }
 
@@ -568,7 +568,7 @@
         this.renderOutlines = renderOutlinesIn;
     }
 
-    public Map<Class, Render> getEntityRenderMap()
+    public Map<Class<? extends Entity>, Render<? extends Entity>> getEntityRenderMap()
     {
         return this.entityRenderMap;
     }
