--- a/net/minecraft/client/renderer/color/BlockColors.java
+++ b/net/minecraft/client/renderer/color/BlockColors.java
@@ -174,10 +174,10 @@
         }
     }
 
-    public int colorMultiplier(IBlockState state, @Nullable IBlockAccess blockAccess, @Nullable BlockPos pos, int RenderPass)
+    public int colorMultiplier(IBlockState state, @Nullable IBlockAccess blockAccess, @Nullable BlockPos pos, int renderPass)
     {
         IBlockColor iblockcolor = this.mapBlockColors.getByValue(Block.getIdFromBlock(state.getBlock()));
-        return iblockcolor == null ? -1 : iblockcolor.colorMultiplier(state, blockAccess, pos, RenderPass);
+        return iblockcolor == null ? -1 : iblockcolor.colorMultiplier(state, blockAccess, pos, renderPass);
     }
 
     public void registerBlockColorHandler(IBlockColor blockColor, Block... blocksIn)
