--- a/net/minecraft/src/ConnectedParser.java
+++ b/net/minecraft/src/ConnectedParser.java
@@ -427,8 +427,9 @@
 
     public static Comparable getPropertyValue(String p_getPropertyValue_0_, Collection p_getPropertyValue_1_)
     {
-        for (Comparable comparable : p_getPropertyValue_1_)
+        for (Object c : p_getPropertyValue_1_)
         {
+        	Comparable comparable = (Comparable) c;
             if (getValueName(comparable).equals(p_getPropertyValue_0_))
             {
                 return comparable;
