--- a/net/minecraft/src/GuiMessage.java
+++ b/net/minecraft/src/GuiMessage.java
@@ -53,8 +53,10 @@
         this.drawCenteredString(this.fontRenderer, this.messageLine1, this.width / 2, 70, 16777215);
         int i = 90;
 
-        for (String s : this.listLines2)
+        for (Object s0 : this.listLines2)
         {
+        	String s = (String) s0;
+
             this.drawCenteredString(this.fontRenderer, s, this.width / 2, i, 16777215);
             i += this.fontRenderer.FONT_HEIGHT;
         }
