--- a/net/minecraft/client/ClientBrandRetriever.java
+++ b/net/minecraft/client/ClientBrandRetriever.java
@@ -1,9 +1,14 @@
 package net.minecraft.client;
 
+import net.optifine.reflect.Reflector;
+
 public class ClientBrandRetriever
 {
     public static String getClientModName()
     {
-        return "vanilla";
+    	if (Reflector.forgeExists()) {
+    		return "vivecraft," + (String)Reflector.call(Reflector.call(Reflector.FMLCommonHandler_instance), Reflector.FMLCommonHandler_getModName);
+    	}
+    	return "vivecraft";
     }
 }
