--- a/net/minecraft/client/LoadingScreenRenderer.java
+++ b/net/minecraft/client/LoadingScreenRenderer.java
@@ -226,7 +226,7 @@
 
                 if (OpenGlHelper.isFramebufferEnabled())
                 {
-                    this.framebuffer.framebufferRender(k * j, l * j);
+                    this.framebuffer.framebufferRender(mc.displayWidth, mc.displayHeight);
                 }
 
                 this.mc.updateDisplay();
