--- a/net/minecraft/client/LoadingScreenRenderer.java
+++ b/net/minecraft/client/LoadingScreenRenderer.java
@@ -8,6 +8,7 @@
 import net.minecraft.client.renderer.Tessellator;
 import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
 import net.minecraft.client.shader.Framebuffer;
+import net.minecraft.src.Reflector;
 import net.minecraft.util.IProgressUpdate;
 import net.minecraft.util.MinecraftError;
 
