--- a/net/minecraft/client/LoadingScreenRenderer.java
+++ b/net/minecraft/client/LoadingScreenRenderer.java
@@ -1,5 +1,7 @@
 package net.minecraft.client;
 
+import com.mtbs3d.minecrift.gameplay.screenhandlers.GuiHandler;
+
 import net.minecraft.client.gui.Gui;
 import net.minecraft.client.gui.ScaledResolution;
 import net.minecraft.client.renderer.BufferBuilder;
@@ -226,7 +228,7 @@
 
                 if (OpenGlHelper.isFramebufferEnabled())
                 {
-                    this.framebuffer.framebufferRender(k * j, l * j);
+                    this.framebuffer.framebufferRender(mc.displayWidth, mc.displayHeight);
                 }
 
                 this.mc.updateDisplay();
