--- a/net/optifine/http/HttpPipeline.java
+++ b/net/optifine/http/HttpPipeline.java
@@ -3,10 +3,12 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InterruptedIOException;
+import java.lang.reflect.Field;
 import java.net.Proxy;
 import java.net.URL;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
 import net.minecraft.src.Config;
 
@@ -212,7 +214,7 @@
 
     public static boolean hasActiveRequests()
     {
-        for (HttpPipelineConnection httppipelineconnection : mapConnections.values())
+        for (HttpPipelineConnection httppipelineconnection : (List<HttpPipelineConnection>)(List<?>)mapConnections.values())
         {
             if (httppipelineconnection.hasActiveRequests())
             {
