--- a/net/minecraft/client/renderer/texture/TextureUtil.java
+++ b/net/minecraft/client/renderer/texture/TextureUtil.java
@@ -272,8 +272,7 @@
     public static int[] readImageData(IResourceManager resourceManager, ResourceLocation imageLocation) throws IOException
     {
         IResource iresource = null;
-        Object i;
-
+        int[] out;
         try
         {
             iresource = resourceManager.getResource(imageLocation);
@@ -289,14 +288,14 @@
                 return aint;
             }
 
-            i = null;
+            out = null;
         }
         finally
         {
             IOUtils.closeQuietly((Closeable)iresource);
         }
 
-        return (int[])i;
+        return out;
     }
 
     public static BufferedImage readBufferedImage(InputStream imageStream) throws IOException
