--- a/net/minecraft/client/renderer/texture/TextureMap.java
+++ b/net/minecraft/client/renderer/texture/TextureMap.java
@@ -109,7 +109,9 @@
         this.missingImage.setIconHeight(i);
         int[][] aint1 = new int[this.mipmapLevels + 1][];
         aint1[0] = aint;
-        this.missingImage.setFramesTextureData(Lists.newArrayList(aint1));
+        ArrayList<int[][]> x = new ArrayList<int[][]>();
+        x.add(aint1);
+        this.missingImage.setFramesTextureData(x);
         this.missingImage.setIndexInMap(this.counterIndexInMap.nextValue());
     }
 
@@ -831,7 +833,7 @@
     {
         Map map1 = new HashMap();
 
-        for (Entry entry : p_detectMinimumSpriteSize_1_.entrySet())
+        for (Entry entry : (Set<Entry>)(Set<?>)p_detectMinimumSpriteSize_1_.entrySet())
         {
             TextureAtlasSprite textureatlassprite6 = (TextureAtlasSprite)entry.getValue();
             ResourceLocation resourcelocation1 = new ResourceLocation(textureatlassprite6.getIconName());
