--- a/net/minecraft/client/renderer/texture/TextureUtil.java
+++ b/net/minecraft/client/renderer/texture/TextureUtil.java
@@ -282,9 +282,9 @@
             if (bufferedimage != null)
             {
                 int j = bufferedimage.getWidth();
-                int i = bufferedimage.getHeight();
-                int[] aint1 = new int[j * i];
-                bufferedimage.getRGB(0, 0, j, i, aint1, 0, j);
+                int h = bufferedimage.getHeight();
+                int[] aint1 = new int[j * h];
+                bufferedimage.getRGB(0, 0, j, h, aint1, 0, j);
                 int[] aint = aint1;
                 return aint;
             }
