--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -20,11 +20,17 @@
 
 public class Main
 {
+	public static boolean kiosk;
+	public static boolean viewonly;
+	public static boolean katvr;
     public static void main(String[] p_main_0_)
     {
         OptionParser optionparser = new OptionParser();
         optionparser.allowsUnrecognizedOptions();
         optionparser.accepts("demo");
+        optionparser.accepts("kiosk");
+        optionparser.accepts("viewonly");
+        optionparser.accepts("katvr");
         optionparser.accepts("fullscreen");
         optionparser.accepts("checkGlErrors");
         OptionSpec<String> optionspec = optionparser.accepts("server").withRequiredArg();
@@ -90,6 +96,14 @@
         boolean flag = optionset.has("fullscreen");
         boolean flag1 = optionset.has("checkGlErrors");
         boolean flag2 = optionset.has("demo");
+        kiosk = optionset.has("kiosk");
+    	if(kiosk) System.out.println("Setting kiosk");
+        if (kiosk) {
+        	viewonly = optionset.has("viewonly");
+        	if(viewonly) 
+        		System.out.println("Setting viewonly");
+        }
+        katvr = optionset.has("katvr");
         String s3 = (String)optionset.valueOf(optionspec12);
         Gson gson = (new GsonBuilder()).registerTypeAdapter(PropertyMap.class, new Serializer()).create();
         PropertyMap propertymap = (PropertyMap)JsonUtils.gsonDeserialize(gson, (String)optionset.valueOf(optionspec15), PropertyMap.class);
