--- a/optifine/Utils.java
+++ b/optifine/Utils.java
@@ -36,14 +36,14 @@
         String s = System.getProperty("user.home", ".");
         File file1 = null;
 
-        switch ($SWITCH_TABLE$optifine$Utils$OS()[getPlatform().ordinal()])
+        switch (getPlatform())
         {
-            case 1:
-            case 2:
+            case LINUX:
+            case SOLARIS:
                 file1 = new File(s, '.' + applicationName + '/');
                 break;
 
-            case 3:
+            case WINDOWS:
                 String s1 = System.getenv("APPDATA");
 
                 if (s1 != null)
@@ -57,7 +57,7 @@
 
                 break;
 
-            case 4:
+            case MACOS:
                 file1 = new File(s, "Library/Application Support/" + applicationName);
                 break;
 
@@ -362,7 +362,7 @@
 
                     if (object instanceof Object[])
                     {
-                        Object[] aobject = object;
+                        Object[] aobject = (Object[]) object;
                         stringbuffer.append(arrayToCommaSeparatedString(aobject));
                     }
                     else
