--- a/net/minecraft/client/renderer/chunk/RenderChunk.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunk.java
@@ -180,8 +180,9 @@
                 boolean flag = Reflector.ForgeBlock_canRenderInLayer.exists();
                 boolean flag1 = Reflector.ForgeHooksClient_setRenderLayer.exists();
 
-                for (BlockPosM blockposm : BlockPosM.getAllInBoxMutable(blockpos, blockpos1))
+                for (Object b : BlockPosM.getAllInBoxMutable(blockpos, blockpos1))
                 {
+                	BlockPosM blockposm= (BlockPosM) b;
                     IBlockState iblockstate = chunkcacheof.getBlockState(blockposm);
                     Block block = iblockstate.getBlock();
 
