--- a/net/minecraft/client/renderer/tileentity/TileEntityEndPortalRenderer.java
+++ b/net/minecraft/client/renderer/tileentity/TileEntityEndPortalRenderer.java
@@ -81,7 +81,7 @@
                 GlStateManager.translate(17.0F / f2, (2.0F + f2 / 1.5F) * ((float)Minecraft.getSystemTime() % 800000.0F / 800000.0F), 0.0F);
                 GlStateManager.rotate((f2 * f2 * 4321.0F + f2 * 9.0F) * 2.0F, 0.0F, 0.0F, 1.0F);
                 GlStateManager.scale(4.5F - f2 / 4.0F, 4.5F - f2 / 4.0F, 1.0F);
-                GlStateManager.multMatrix(PROJECTION);
+                //GlStateManager.multMatrix(PROJECTION);
                 GlStateManager.multMatrix(MODELVIEW);
                 Tessellator tessellator = Tessellator.getInstance();
                 BufferBuilder bufferbuilder = tessellator.getBuffer();
