--- a/net/minecraft/src/ConnectedParser.java
+++ b/net/minecraft/src/ConnectedParser.java
@@ -427,7 +427,7 @@
 
     public static Comparable getPropertyValue(String p_getPropertyValue_0_, Collection p_getPropertyValue_1_)
     {
-        for (Comparable comparable : p_getPropertyValue_1_)
+        for (Comparable comparable : (Iterable<Comparable>)p_getPropertyValue_1_)
         {
             if (getValueName(comparable).equals(p_getPropertyValue_0_))
             {
