--- a/net/minecraft/src/CustomGuiProperties.java
+++ b/net/minecraft/src/CustomGuiProperties.java
@@ -3,6 +3,8 @@
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Set;
+
 import net.minecraft.block.BlockChest;
 import net.minecraft.client.gui.GuiEnchantment;
 import net.minecraft.client.gui.GuiHopper;
@@ -199,7 +201,7 @@
 
         String s5 = p_parseTextureLocations_1_ + ".";
 
-        for (String s1 : p_parseTextureLocations_0_.keySet())
+        for (String s1 : (Set<String>)(Set<?>)p_parseTextureLocations_0_.keySet())
         {
             if (s1.startsWith(s5))
             {
