--- a/net/minecraft/src/CustomColors.java
+++ b/net/minecraft/src/CustomColors.java
@@ -318,7 +318,7 @@
 
         if (ainteger.length <= 0)
         {
-            return new ImmutablePair<CustomColormap[], Integer>((Object)null, Integer.valueOf(0));
+            return new ImmutablePair<CustomColormap[], Integer>((CustomColormap[])null, Integer.valueOf(0));
         }
         else
         {
@@ -436,8 +436,9 @@
         String s = "palette.block.";
         Map map = new HashMap();
 
-        for (String s1 : p_readCustomColormaps_0_.keySet())
+        for (Object s0 : p_readCustomColormaps_0_.keySet())
         {
+        	String s1 = (String) s0;
             String s2 = p_readCustomColormaps_0_.getProperty(s1);
 
             if (s1.startsWith(s))
@@ -1453,8 +1454,9 @@
         Set set = p_readSpawnEggColors_0_.keySet();
         int i = 0;
 
-        for (String s : set)
+        for (Object s0 : set)
         {
+        	String s = (String) s0;
             String s1 = p_readSpawnEggColors_0_.getProperty(s);
 
             if (s.startsWith(p_readSpawnEggColors_2_))
@@ -1592,8 +1594,9 @@
         float[][] afloat1 = new float[aenumdyecolor.length][];
         int k = 0;
 
-        for (String s : p_readDyeColors_0_.keySet())
+        for (Object s0 : p_readDyeColors_0_.keySet())
         {
+        	String s = (String) s0;
             String s1 = p_readDyeColors_0_.getProperty(s);
 
             if (s.startsWith(p_readDyeColors_2_))
@@ -1665,8 +1668,9 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : p_readTextColors_0_.keySet())
+        for (Object s0 : p_readTextColors_0_.keySet())
         {
+        	String s = (String) s0;
             String s1 = p_readTextColors_0_.getProperty(s);
 
             if (s.startsWith(p_readTextColors_2_))
@@ -1721,8 +1725,9 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : p_readMapColors_0_.keySet())
+        for (Object s0 : p_readMapColors_0_.keySet())
         {
+        	String s = (String) s0;
             String s1 = p_readMapColors_0_.getProperty(s);
 
             if (s.startsWith(p_readMapColors_2_))
@@ -1760,8 +1765,9 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : p_readPotionColors_0_.keySet())
+        for (Object s0 : p_readPotionColors_0_.keySet())
         {
+        	String s = (String) s0;
             String s1 = p_readPotionColors_0_.getProperty(s);
 
             if (s.startsWith(p_readPotionColors_2_))
