--- a/net/minecraft/src/CustomColors.java
+++ b/net/minecraft/src/CustomColors.java
@@ -321,7 +321,7 @@
 
         if (ainteger.length <= 0)
         {
-            return new ImmutablePair<LightMapPack[], Integer>((Object)null, Integer.valueOf(0));
+            return new ImmutablePair<LightMapPack[], Integer>((LightMapPack[])null, Integer.valueOf(0));
         }
         else
         {
@@ -459,7 +459,7 @@
         String s = "palette.block.";
         Map map = new HashMap();
 
-        for (String s1 : p_readCustomColormaps_0_.keySet())
+        for (String s1 : (Set<String>)(Set<?>)p_readCustomColormaps_0_.keySet())
         {
             String s2 = p_readCustomColormaps_0_.getProperty(s1);
 
@@ -1372,7 +1372,7 @@
         Set set = p_readSpawnEggColors_0_.keySet();
         int i = 0;
 
-        for (String s : set)
+        for (String s : (Set<String>)(Set<?>)set)
         {
             String s1 = p_readSpawnEggColors_0_.getProperty(s);
 
@@ -1511,7 +1511,7 @@
         float[][] afloat1 = new float[aenumdyecolor.length][];
         int k = 0;
 
-        for (String s : p_readDyeColors_0_.keySet())
+        for (String s : (Set<String>)(Set<?>)p_readDyeColors_0_.keySet())
         {
             String s1 = p_readDyeColors_0_.getProperty(s);
 
@@ -1584,7 +1584,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : p_readTextColors_0_.keySet())
+        for (String s : (Set<String>)(Set<?>)p_readTextColors_0_.keySet())
         {
             String s1 = p_readTextColors_0_.getProperty(s);
 
@@ -1640,7 +1640,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : p_readMapColors_0_.keySet())
+        for (String s : (Set<String>)(Set<?>)p_readMapColors_0_.keySet())
         {
             String s1 = p_readMapColors_0_.getProperty(s);
 
@@ -1679,7 +1679,7 @@
         Arrays.fill(aint, -1);
         int i = 0;
 
-        for (String s : p_readPotionColors_0_.keySet())
+        for (String s : (Set<String>)(Set<?>) p_readPotionColors_0_.keySet())
         {
             String s1 = p_readPotionColors_0_.getProperty(s);
 
