--- a/net/minecraft/src/FontUtils.java
+++ b/net/minecraft/src/FontUtils.java
@@ -50,8 +50,10 @@
 
     public static void readCustomCharWidths(Properties p_readCustomCharWidths_0_, float[] p_readCustomCharWidths_1_)
     {
-        for (String s : p_readCustomCharWidths_0_.keySet())
+        for (Object s0 : p_readCustomCharWidths_0_.keySet())
         {
+        	String s = (String) s0;
+
             String s1 = "width.";
 
             if (s.startsWith(s1))
