--- a/net/minecraft/src/GuiMessage.java
+++ b/net/minecraft/src/GuiMessage.java
@@ -53,7 +53,7 @@
         this.drawCenteredString(this.fontRenderer, this.messageLine1, this.width / 2, 70, 16777215);
         int i = 90;
 
-        for (String s : this.listLines2)
+        for (String s : (List<String>)this.listLines2)
         {
             this.drawCenteredString(this.fontRenderer, s, this.width / 2, i, 16777215);
             i += this.fontRenderer.FONT_HEIGHT;
