--- a/net/minecraft/src/HttpPipeline.java
+++ b/net/minecraft/src/HttpPipeline.java
@@ -211,7 +211,7 @@
 
     public static boolean hasActiveRequests()
     {
-        for (HttpPipelineConnection httppipelineconnection : mapConnections.values())
+        for (HttpPipelineConnection httppipelineconnection : (Iterable<HttpPipelineConnection>)mapConnections.values())
         {
             if (httppipelineconnection.hasActiveRequests())
             {
