--- a/net/minecraft/src/HttpPipeline.java
+++ b/net/minecraft/src/HttpPipeline.java
@@ -200,8 +200,9 @@
 
     public static boolean hasActiveRequests()
     {
-        for (HttpPipelineConnection httppipelineconnection : mapConnections.values())
+        for (Object h : mapConnections.values())
         {
+        	HttpPipelineConnection httppipelineconnection = (HttpPipelineConnection) h;
             if (httppipelineconnection.hasActiveRequests())
             {
                 return true;
