--- a/net/minecraft/src/HttpUtils.java
+++ b/net/minecraft/src/HttpUtils.java
@@ -9,6 +9,8 @@
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.util.Map;
+import java.util.Set;
+
 import net.minecraft.client.Minecraft;
 
 public class HttpUtils
@@ -87,7 +89,7 @@
 
             if (p_post_1_ != null)
             {
-                for (String s : p_post_1_.keySet())
+                for (String s : (Set<String>)(Set<?>)p_post_1_.keySet())
                 {
                     String s1 = "" + p_post_1_.get(s);
                     httpurlconnection.setRequestProperty(s, s1);
