--- a/net/minecraft/src/ModelUtils.java
+++ b/net/minecraft/src/ModelUtils.java
@@ -33,7 +33,7 @@
 
     private static void dbgQuads(String p_dbgQuads_0_, List p_dbgQuads_1_, String p_dbgQuads_2_)
     {
-        for (BakedQuad bakedquad : p_dbgQuads_1_)
+        for (BakedQuad bakedquad : (List<BakedQuad>)p_dbgQuads_1_)
         {
             dbgQuad(p_dbgQuads_0_, bakedquad, p_dbgQuads_2_);
         }
@@ -85,7 +85,7 @@
     {
         List list = new ArrayList();
 
-        for (BakedQuad bakedquad : p_duplicateQuadList_0_)
+        for (BakedQuad bakedquad : (Iterable<BakedQuad>)p_duplicateQuadList_0_)
         {
             BakedQuad bakedquad1 = duplicateQuad(bakedquad);
             list.add(bakedquad1);
