--- a/net/minecraft/src/RandomMobs.java
+++ b/net/minecraft/src/RandomMobs.java
@@ -68,7 +68,7 @@
         }
         else
         {
-            ResourceLocation entity;
+            ResourceLocation out;
 
             try
             {
@@ -110,14 +110,14 @@
                     return resourcelocation1;
                 }
 
-                entity = p_getTextureLocation_0_;
+                out = p_getTextureLocation_0_;
             }
             finally
             {
                 working = false;
             }
 
-            return entity;
+            return out;
         }
     }
 
