--- a/net/minecraft/src/SmartLeaves.java
+++ b/net/minecraft/src/SmartLeaves.java
@@ -174,7 +174,7 @@
                         }
                         else
                         {
-                            for (BakedQuad bakedquad : list)
+                            for (BakedQuad bakedquad : (List<BakedQuad>)list)
                             {
                                 List list1 = ibakedmodel.getQuads((IBlockState)null, bakedquad.getFace(), 0L);
 
