--- a/net/minecraft/block/state/BlockStateBase.java
+++ b/net/minecraft/block/state/BlockStateBase.java
@@ -32,7 +32,7 @@
         }
         private <T extends Comparable<T>> String getPropertyName(IProperty<T> property, Comparable<?> entry)
         {
-            return property.getName(entry);
+            return property.getName((T) entry);
         }
     };
     private int blockId = -1;
@@ -87,7 +87,7 @@
 
     public <T extends Comparable<T>> IBlockState cycleProperty(IProperty<T> property)
     {
-        return this.withProperty(property, (Comparable)cyclePropertyValue(property.getAllowedValues(), this.getValue(property)));
+        return this.withProperty(property, cyclePropertyValue(property.getAllowedValues(), this.getValue(property)));
     }
 
     protected static <T> T cyclePropertyValue(Collection<T> values, T currentValue)
