--- a/net/optifine/ConnectedProperties.java
+++ b/net/optifine/ConnectedProperties.java
@@ -1,5 +1,6 @@
 package net.optifine;
 
+import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
@@ -473,7 +474,7 @@
 
     public static IProperty getProperty(String key, Collection properties)
     {
-        for (IProperty iproperty : properties)
+        for (IProperty iproperty : (Collection<IProperty>)(Collection<?>)properties)
         {
             if (key.equals(iproperty.getName()))
             {
