--- a/net/optifine/ConnectedProperties.java
+++ b/net/optifine/ConnectedProperties.java
@@ -6,6 +6,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Set;
 import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.client.Minecraft;
@@ -473,7 +474,7 @@
 
     public static IProperty getProperty(String key, Collection properties)
     {
-        for (IProperty iproperty : properties)
+        for (IProperty iproperty : (Set<IProperty>)(Set<?>)properties)
         {
             if (key.equals(iproperty.getName()))
             {
