--- a/net/optifine/ConnectedTextures.java
+++ b/net/optifine/ConnectedTextures.java
@@ -2339,13 +2339,13 @@
         return !set1.isEmpty();
     }
 
-    private static ConnectedProperties[][] propertyListToArray(List list)
+    private static ConnectedProperties[][] propertyListToArray(List listp)
     {
-        ConnectedProperties[][] aconnectedproperties = new ConnectedProperties[list.size()][];
+        ConnectedProperties[][] aconnectedproperties = new ConnectedProperties[listp.size()][];
 
-        for (int i = 0; i < list.size(); ++i)
+        for (int i = 0; i < listp.size(); ++i)
         {
-            List list = (List)list.get(i);
+            List list = (List)listp.get(i);
 
             if (list != null)
             {
@@ -2413,15 +2413,15 @@
             list.add((Object)null);
         }
 
-        List list = (List)list.get(id);
+        List sublist = (List)list.get(id);
 
-        if (list == null)
+        if (sublist == null)
         {
-            list = new ArrayList();
-            list.set(id, list);
+            sublist = new ArrayList();
+            list.set(id, sublist);
         }
 
-        list.add(cp);
+        sublist.add(cp);
     }
 
     private static String[] getDefaultCtmPaths()
