--- a/net/optifine/CustomGuiProperties.java
+++ b/net/optifine/CustomGuiProperties.java
@@ -3,6 +3,8 @@
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Set;
+
 import net.minecraft.block.BlockChest;
 import net.minecraft.client.gui.GuiEnchantment;
 import net.minecraft.client.gui.GuiHopper;
@@ -194,12 +196,12 @@
     private static Map<ResourceLocation, ResourceLocation> parseTextureLocations(Properties props, String property, CustomGuiProperties.EnumContainer container, String pathPrefix, String basePath)
     {
         Map<ResourceLocation, ResourceLocation> map = new HashMap<ResourceLocation, ResourceLocation>();
-        String s = props.getProperty(property);
+        String propval = props.getProperty(property);
 
-        if (s != null)
+        if (propval != null)
         {
             ResourceLocation resourcelocation = getGuiTextureLocation(container);
-            ResourceLocation resourcelocation1 = parseTextureLocation(s, basePath);
+            ResourceLocation resourcelocation1 = parseTextureLocation(propval, basePath);
 
             if (resourcelocation != null && resourcelocation1 != null)
             {
@@ -209,7 +211,7 @@
 
         String s5 = property + ".";
 
-        for (String s1 : props.keySet())
+        for (String s1 : (Set<String>)(Set<?>)props.keySet())
         {
             if (s1.startsWith(s5))
             {
