--- a/net/optifine/Mipmaps.java
+++ b/net/optifine/Mipmaps.java
@@ -171,10 +171,10 @@
         return k << 24 | j2 << 16 | k2 << 8 | l2;
     }
 
-    private int averageColor(int i, int j)
+    private int averageColor(int ii, int jj)
     {
-        int i = (i & -16777216) >> 24 & 255;
-        int j = (j & -16777216) >> 24 & 255;
+        int i = (ii & -16777216) >> 24 & 255;
+        int j = (jj & -16777216) >> 24 & 255;
         return (i + j >> 1 << 24) + ((i & 16711422) + (j & 16711422) >> 1);
     }
 
