--- a/net/optifine/SmartLeaves.java
+++ b/net/optifine/SmartLeaves.java
@@ -176,7 +176,7 @@
                         }
                         else
                         {
-                            for (BakedQuad bakedquad : list)
+                            for (BakedQuad bakedquad : (List<BakedQuad>)list)
                             {
                                 List list1 = ibakedmodel.getQuads((IBlockState)null, bakedquad.getFace(), 0L);
 
