--- a/net/minecraft/client/gui/inventory/GuiEditSign.java
+++ b/net/minecraft/client/gui/inventory/GuiEditSign.java
@@ -1,7 +1,9 @@
 package net.minecraft.client.gui.inventory;
 
 import java.io.IOException;
+
 import net.minecraft.block.Block;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.GuiButton;
 import net.minecraft.client.gui.GuiScreen;
 import net.minecraft.client.network.NetHandlerPlayClient;
@@ -13,8 +15,12 @@
 import net.minecraft.tileentity.TileEntitySign;
 import net.minecraft.util.ChatAllowedCharacters;
 import net.minecraft.util.text.TextComponentString;
+
 import org.lwjgl.input.Keyboard;
 
+import com.mtbs3d.minecrift.gameplay.screenhandlers.KeyboardHandler;
+import com.mtbs3d.minecrift.provider.MCOpenVR;
+
 public class GuiEditSign extends GuiScreen
 {
     /** Reference to the sign object. */
@@ -44,6 +50,7 @@
         Keyboard.enableRepeatEvents(true);
         this.doneBtn = this.addButton(new GuiButton(0, this.width / 2 - 100, this.height / 4 + 120, I18n.format("gui.done")));
         this.tileSign.setEditable(false);
+       	KeyboardHandler.setOverlayShowing(true);
     }
 
     /**
@@ -60,6 +67,7 @@
         }
 
         this.tileSign.setEditable(true);
+        KeyboardHandler.setOverlayShowing(false);
     }
 
     /**
