--- a/net/minecraft/block/state/BlockStateContainer.java
+++ b/net/minecraft/block/state/BlockStateContainer.java
@@ -562,10 +562,10 @@
         {
             return Reflector.callBoolean(this.block, Reflector.ForgeBlock_isSideSolid, this, p_isSideSolid_1_, p_isSideSolid_2_, p_isSideSolid_3_);
         }
-
-        public boolean doesSideBlockChestOpening(IBlockAccess p_doesSideBlockChestOpening_1_, BlockPos p_doesSideBlockChestOpening_2_, EnumFacing p_doesSideBlockChestOpening_3_)
+        
+        public boolean doesSideBlockChestOpening(IBlockAccess p_isSideSolid_1_, BlockPos p_isSideSolid_2_, EnumFacing p_isSideSolid_3_)
         {
-            return Reflector.callBoolean(this.block, Reflector.ForgeBlock_doesSideBlockChestOpening, this, p_doesSideBlockChestOpening_1_, p_doesSideBlockChestOpening_2_, p_doesSideBlockChestOpening_3_);
+            return Reflector.callBoolean(this.block, Reflector.ForgeBlock_isSideSolid, this, p_isSideSolid_1_, p_isSideSolid_2_, p_isSideSolid_3_);
         }
 
         public boolean doesSideBlockRendering(IBlockAccess p_doesSideBlockRendering_1_, BlockPos p_doesSideBlockRendering_2_, EnumFacing p_doesSideBlockRendering_3_)
