--- a/net/minecraft/client/renderer/entity/RenderLivingBase.java
+++ b/net/minecraft/client/renderer/entity/RenderLivingBase.java
@@ -58,7 +58,7 @@
 
     public <V extends EntityLivingBase, U extends LayerRenderer<V>> boolean addLayer(U layer)
     {
-        return this.layerRenderers.add(layer);
+        return this.layerRenderers.add((LayerRenderer<T>) layer);
     }
 
     public ModelBase getMainModel()
@@ -629,6 +629,39 @@
 
     protected boolean canRenderName(T entity)
     {
+        EntityPlayerSP entityplayersp = Minecraft.getMinecraft().player;
+        boolean flag = !entity.isInvisibleToPlayer(entityplayersp);
+
+        if (entity != entityplayersp)
+        {
+            Team team = entity.getTeam();
+            Team team1 = entityplayersp.getTeam();
+
+            if (team != null)
+            {
+                Team.EnumVisible team$enumvisible = team.getNameTagVisibility();
+
+                switch (team$enumvisible)
+                {
+                    case ALWAYS:
+                        return flag;
+
+                    case NEVER:
+                        return false;
+
+                    case HIDE_FOR_OTHER_TEAMS:
+                        return team1 == null ? flag : team.isSameTeam(team1) && (team.getSeeFriendlyInvisiblesEnabled() || flag);
+
+                    case HIDE_FOR_OWN_TEAM:
+                        return team1 == null ? flag : !team.isSameTeam(team1) && flag;
+
+                    default:
+                        return true;
+                }
+            }
+        }
+
+        return Minecraft.isGuiEnabled() && entity != this.renderManager.renderViewEntity && flag && !entity.isBeingRidden();
     }
 
     public List<LayerRenderer<T>> getLayerRenderers()
