--- a/net/minecraft/client/renderer/ActiveRenderInfo.java
+++ b/net/minecraft/client/renderer/ActiveRenderInfo.java
@@ -4,6 +4,7 @@
 import java.nio.IntBuffer;
 import net.minecraft.block.BlockLiquid;
 import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.Minecraft;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.util.math.BlockPos;
@@ -68,6 +69,7 @@
         rotationXZ = MathHelper.cos(f2 * 0.017453292F);
     }
 
+    //unused now.
     public static Vec3d projectViewFromEntity(Entity entityIn, double p_178806_1_)
     {
         double d0 = entityIn.prevPosX + (entityIn.posX - entityIn.prevPosX) * p_178806_1_;
@@ -79,6 +81,7 @@
         return new Vec3d(d3, d4, d5);
     }
 
+    //unused now.
     public static IBlockState getBlockStateAtEntityViewpoint(World worldIn, Entity entityIn, float p_186703_2_)
     {
         Vec3d vec3d = projectViewFromEntity(entityIn, (double)p_186703_2_);
@@ -105,6 +108,8 @@
         return iblockstate;
     }
 
+    //These are used ONLY for particles.vvv
+    
     public static float getRotationX()
     {
         return rotationX;
