--- a/net/minecraft/client/renderer/GlStateManager.java
+++ b/net/minecraft/client/renderer/GlStateManager.java
@@ -21,7 +21,7 @@
     private static final GlStateManager.BooleanState lightingState = new GlStateManager.BooleanState(2896);
     private static final GlStateManager.BooleanState[] lightState = new GlStateManager.BooleanState[8];
     private static final GlStateManager.ColorMaterialState colorMaterialState;
-    private static final GlStateManager.BlendState blendState;
+    public static final GlStateManager.BlendState blendState;
     private static final GlStateManager.DepthState depthState;
     private static final GlStateManager.FogState fogState;
     private static final GlStateManager.CullState cullState;
@@ -394,7 +394,23 @@
 
     private static GlStateManager.TexGenCoord texGenCoord(GlStateManager.TexGen texGen)
     {
-		return null;
+        switch (texGen)
+        {
+            case S:
+                return texGenState.s;
+
+            case T:
+                return texGenState.t;
+
+            case R:
+                return texGenState.r;
+
+            case Q:
+                return texGenState.q;
+
+            default:
+                return texGenState.s;
+        }
     }
 
     public static void setActiveTexture(int texture)
@@ -1003,7 +1019,7 @@
         }
     }
 
-    static class Color
+    public static class Color
     {
         public float red;
         public float green;
