--- a/net/minecraft/world/Teleporter.java
+++ b/net/minecraft/world/Teleporter.java
@@ -14,8 +14,10 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.math.MathHelper;
+import net.minecraftforge.common.util.ITeleporter;
 
-public class Teleporter
+//FORGE: implement ITeleporter
+public class Teleporter implements ITeleporter
 {
     private final WorldServer world;
 
@@ -434,4 +436,15 @@
             this.lastUpdateTime = lastUpdate;
         }
     }
+
+    //FORGE
+    @Override
+    public void placeEntity(World world, Entity entity, float yaw)
+    {
+        if (entity instanceof EntityPlayerMP)
+            placeInPortal(entity, yaw);
+        else
+            placeInExistingPortal(entity, yaw);
+    }
+    //
 }
