--- a/net/minecraft/world/WorldEntitySpawner.java
+++ b/net/minecraft/world/WorldEntitySpawner.java
@@ -229,7 +229,7 @@
                                                         entityliving.setDead();
                                                     }
 
-                                                    int j4 = Reflector.ForgeEventFactory_getMaxSpawnedInChunk.exists() ? Reflector.callInt(Reflector.ForgeEventFactory_getMaxSpawnedInChunk, entityliving) : entityliving.getMaxSpawnedInChunk();
+                                                    int j4 = Reflector.ForgeEventFactory_getMaxSpawnPackSize.exists() ? Reflector.callInt(Reflector.ForgeEventFactory_getMaxSpawnPackSize, entityliving) : entityliving.getMaxSpawnedInChunk();
 
                                                     if (j2 >= j4)
                                                     {
