--- a/net/minecraft/client/renderer/entity/Render.java
+++ b/net/minecraft/client/renderer/entity/Render.java
@@ -23,6 +23,7 @@
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 import net.optifine.entity.model.IEntityRenderer;
 import net.optifine.shaders.Shaders;
@@ -392,18 +393,20 @@
      */
     protected void renderLivingLabel(T entityIn, String str, double x, double y, double z, int maxDistance)
     {
-        double d0 = entityIn.getDistanceSqToEntity(this.renderManager.renderViewEntity);
+    	double d0 = entityIn.getDistanceSqToEntity(this.renderManager.renderViewEntity);
 
-        if (d0 <= (double)(maxDistance * maxDistance))
-        {
-            boolean flag = entityIn.isSneaking();
-            float f = this.renderManager.playerViewY;
-            float f1 = this.renderManager.playerViewX;
-            boolean flag1 = this.renderManager.options.thirdPersonView == 2;
-            float f2 = entityIn.height + 0.5F - (flag ? 0.25F : 0.0F);
-            int i = "deadmau5".equals(str) ? -10 : 0;
-            EntityRenderer.drawNameplate(this.getFontRendererFromRenderManager(), str, (float)x, (float)y + f2, (float)z, i, f, f1, flag1, flag);
-        }
+    	if (d0 <= (double)(maxDistance * maxDistance))
+    	{
+    		//VIVE redo to fix nameplates being not right
+    		boolean flag = entityIn.isSneaking();
+    		boolean flag1 = this.renderManager.options.thirdPersonView == 2;
+    		float f2 = entityIn.height + 0.5F - (flag ? 0.25F : 0.0F);
+    		y=y+f2;
+    		float f = (float)Math.toDegrees(Math.atan2(-x, z));     
+    		float f1 = -(float)Math.toDegrees(Math.asin(y/Math.sqrt(x*x+y*y+z*z))); 
+    		int i = "deadmau5".equals(str) ? -10 : 0;
+    		EntityRenderer.drawNameplate(this.getFontRendererFromRenderManager(), str, (float)x, (float)y, (float)z, i, f, f1, flag1, flag);
+    	}
     }
 
     public RenderManager getRenderManager()
