--- a/net/minecraft/client/renderer/entity/RenderLivingBase.java
+++ b/net/minecraft/client/renderer/entity/RenderLivingBase.java
@@ -1,11 +1,16 @@
 package net.minecraft.client.renderer.entity;
 
 import com.google.common.collect.Lists;
+import com.mtbs3d.minecrift.render.PlayerModelController;
+
 import java.nio.FloatBuffer;
 import java.util.List;
+import java.util.UUID;
+
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.client.model.ModelBase;
+import net.minecraft.client.model.ModelPlayerVR;
 import net.minecraft.client.model.ModelSpider;
 import net.minecraft.client.renderer.GLAllocation;
 import net.minecraft.client.renderer.GlStateManager;
@@ -18,6 +23,7 @@
 import net.minecraft.scoreboard.Team;
 import net.minecraft.src.Config;
 import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Vec3d;
 import net.minecraft.util.text.TextFormatting;
 import net.optifine.EmissiveTextures;
 import net.optifine.entity.model.CustomEntityModels;
@@ -48,6 +54,7 @@
     private boolean renderLayersPushMatrix;
     public static final boolean animateModelLiving = Boolean.getBoolean("animate.model.living");
 
+    
     public RenderLivingBase(RenderManager renderManagerIn, ModelBase modelBaseIn, float shadowSizeIn)
     {
         super(renderManagerIn);
@@ -56,6 +63,7 @@
         this.renderModelPushMatrix = this.mainModel instanceof ModelSpider;
     }
 
+    //Forge makes this public but its already public?
     public <V extends EntityLivingBase, U extends LayerRenderer<V>> boolean addLayer(U layer)
     {
         return this.layerRenderers.add((LayerRenderer<T>) layer);
@@ -152,7 +160,29 @@
                 float f7 = entity.prevRotationPitch + (entity.rotationPitch - entity.prevRotationPitch) * partialTicks;
                 this.renderLivingAt(entity, x, y, z);
                 float f8 = this.handleRotationFloat(entity, partialTicks);
+                
+                //VIVECRAFT
+                if(this.mainModel instanceof ModelPlayerVR && entity instanceof EntityPlayer){
+                    UUID uuid = entity.getUniqueID();
+                    ModelPlayerVR mp = (ModelPlayerVR) mainModel;
+                    double d3 = entity.lastTickPosX + (entity.posX - entity.lastTickPosX) * (double)partialTicks;
+                    double d4 = entity.lastTickPosY + (entity.posY - entity.lastTickPosY) * (double)partialTicks;
+                    double d5 = entity.lastTickPosZ + (entity.posZ - entity.lastTickPosZ) * (double)partialTicks;
+                    mp.renderPos = new Vec3d(d3, d4, d5);
+                	if(PlayerModelController.getInstance().isTracked(uuid)){
+                     	PlayerModelController.RotInfo rotInfo=PlayerModelController.getInstance().getRotationsForPlayer(uuid);	
+                		Vec3d diff = rotInfo.leftArmPos.subtract(rotInfo.rightArmPos).rotateYaw((float)-Math.PI/2);
+                		if(rotInfo.reverse) diff = diff.scale(-1);
+                		if(rotInfo.seated) diff = rotInfo.rightArmRot;
+                		double ltor = Math.atan2(-diff.x, diff.z);  
+                		f = (float) Math.toDegrees(ltor);
+                	}
+                }
+
+                float wasyaw = entity.rotationYaw;
+
                 this.applyRotations(entity, f8, f, partialTicks);
+
                 float f4 = this.prepareScale(entity, partialTicks);
                 float f5 = 0.0F;
                 float f6 = 0.0F;
