--- a/shadersmod/client/SVertexFormat.java
+++ b/shadersmod/client/SVertexFormat.java
@@ -6,7 +6,7 @@
 public class SVertexFormat
 {
     public static final int vertexSizeBlock = 14;
-    public static final int offsetMidTexCoord = 8;
+    public static final int offsetMidTex = 8;
     public static final int offsetTangent = 10;
     public static final int offsetEntity = 12;
     public static final VertexFormat defVertexFormatTextured = makeDefVertexFormatTextured();
