--- a/optifine/json/JSONValue.java
+++ b/optifine/json/JSONValue.java
@@ -161,11 +161,11 @@
         }
     }
 
-    static void escape(String s, StringBuffer sb)
+    static void escape(String in, StringBuffer sb)
     {
-        for (int i = 0; i < s.length(); ++i)
+        for (int i = 0; i < in.length(); ++i)
         {
-            char c0 = s.charAt(i);
+            char c0 = in.charAt(i);
 
             switch (c0)
             {
