--- a/net/optifine/model/ModelUtils.java
+++ b/net/optifine/model/ModelUtils.java
@@ -1,5 +1,6 @@
 package net.optifine.model;
 
+import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
@@ -34,7 +35,7 @@
 
     private static void dbgQuads(String name, List quads, String prefix)
     {
-        for (BakedQuad bakedquad : quads)
+        for (BakedQuad bakedquad : (List<BakedQuad>)(List<?>)quads)
         {
             dbgQuad(name, bakedquad, prefix);
         }
@@ -84,15 +85,15 @@
 
     public static List duplicateQuadList(List list)
     {
-        List list = new ArrayList();
+        List sublist = new ArrayList();
 
-        for (BakedQuad bakedquad : list)
+        for (BakedQuad bakedquad : (List<BakedQuad>)(List<?>)list)
         {
             BakedQuad bakedquad1 = duplicateQuad(bakedquad);
-            list.add(bakedquad1);
+            sublist.add(bakedquad1);
         }
 
-        return list;
+        return sublist;
     }
 
     public static BakedQuad duplicateQuad(BakedQuad quad)
