--- a/net/optifine/http/HttpUtils.java
+++ b/net/optifine/http/HttpUtils.java
@@ -90,7 +90,7 @@
 
             if (headers != null)
             {
-                for (String s : headers.keySet())
+                for (String s : (Set<String>)(Set<?>)headers.keySet())
                 {
                     String s1 = "" + headers.get(s);
                     httpurlconnection.setRequestProperty(s, s1);
