--- a/net/optifine/entity/model/anim/ExpressionParser.java
+++ b/net/optifine/entity/model/anim/ExpressionParser.java
@@ -211,7 +211,7 @@
         }
     }
 
-    private EnumFunctionType getFunctionType(Token token, Deque<Token> deque) throws ParseException
+    private EnumFunctionType getFunctionType(Token tokenin, Deque<Token> deque) throws ParseException
     {
         Token token = deque.peek();
 
@@ -223,7 +223,7 @@
         }
         else
         {
-            EnumFunctionType enumfunctiontype = EnumFunctionType.parse(token.getText());
+            EnumFunctionType enumfunctiontype = EnumFunctionType.parse(tokenin.getText());
 
             if (enumfunctiontype == null)
             {
@@ -240,7 +240,7 @@
         }
     }
 
-    private IExpression makeFunction(EnumFunctionType type, Deque<Token> deque) throws ParseException
+    private IExpression makeFunction(EnumFunctionType type, Deque<Token> dequein) throws ParseException
     {
         if (type.getCountArguments() == 0)
         {
@@ -248,20 +248,20 @@
         }
         else
         {
-            Token token = deque.poll();
-            Deque<Token> deque = getGroup(deque, EnumTokenType.BRACKET_CLOSE, true);
+            Token token = dequein.poll();
+            Deque<Token> deque = getGroup(dequein, EnumTokenType.BRACKET_CLOSE, true);
             IExpression[] aiexpression = this.parseExpressions(deque);
             return makeFunction(type, aiexpression);
         }
     }
 
-    private IExpression[] parseExpressions(Deque<Token> deque) throws ParseException
+    private IExpression[] parseExpressions(Deque<Token> dequein) throws ParseException
     {
         List<IExpression> list = new ArrayList<IExpression>();
 
         while (true)
         {
-            Deque<Token> deque = getGroup(deque, EnumTokenType.COMMA, false);
+            Deque<Token> deque = getGroup(dequein, EnumTokenType.COMMA, false);
             IExpression iexpression = this.parseInfix(deque);
 
             if (iexpression == null)
@@ -307,17 +307,17 @@
         }
     }
 
-    private IExpression makeBracketed(Token token, Deque<Token> deque) throws ParseException
+    private IExpression makeBracketed(Token token, Deque<Token> dequein) throws ParseException
     {
-        Deque<Token> deque = getGroup(deque, EnumTokenType.BRACKET_CLOSE, true);
+        Deque<Token> deque = getGroup(dequein, EnumTokenType.BRACKET_CLOSE, true);
         return this.parseInfix(deque);
     }
 
-    private static Deque<Token> getGroup(Deque<Token> deque, EnumTokenType tokenTypeEnd, boolean tokenEndRequired) throws ParseException
+    private static Deque<Token> getGroup(Deque<Token> dequein, EnumTokenType tokenTypeEnd, boolean tokenEndRequired) throws ParseException
     {
         Deque<Token> deque = new ArrayDeque<Token>();
         int i = 0;
-        Iterator iterator = deque.iterator();
+        Iterator iterator = dequein.iterator();
 
         while (iterator.hasNext())
         {
