--- a/net/minecraft/item/ItemFood.java
+++ b/net/minecraft/item/ItemFood.java
@@ -1,6 +1,8 @@
 package net.minecraft.item;
 
 import net.minecraft.advancements.CriteriaTriggers;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.player.EntityPlayer;
@@ -83,6 +85,11 @@
         {
             player.addPotionEffect(new PotionEffect(this.potionId));
         }
+
+		if (player instanceof EntityPlayerSP && stack.getDisplayName().equals("EAT ME")) {
+			Minecraft.getMinecraft().vrPlayer.wfMode = 1;
+			Minecraft.getMinecraft().vrPlayer.wfCount = 400;
+		}
     }
 
     /**
@@ -105,7 +112,7 @@
     {
         ItemStack itemstack = playerIn.getHeldItem(handIn);
 
-        if (playerIn.canEat(this.alwaysEdible))
+        if (playerIn.canEat(this.alwaysEdible) || itemstack.getDisplayName().equals("EAT ME"))
         {
             playerIn.setActiveHand(handIn);
             return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
