--- a/net/optifine/ConnectedProperties.java
+++ b/net/optifine/ConnectedProperties.java
@@ -39,7 +39,7 @@
     public int faces = 63;
     public Biome[] biomes = null;
     public RangeListInt heights = null;
-    public int RenderPass = 0;
+    public int renderPass = 0;
     public boolean innerSeams = false;
     public int[] ctmTileIndexes = null;
     public int width = 0;
@@ -122,7 +122,7 @@
             }
         }
 
-        this.RenderPass = connectedparser.parseInt(props.getProperty("RenderPass"), -1);
+        this.renderPass = connectedparser.parseInt(props.getProperty("renderPass"), -1);
         this.innerSeams = connectedparser.parseBoolean(props.getProperty("innerSeams"), false);
         this.ctmTileIndexes = this.parseCtmTileIndexes(props);
         this.width = connectedparser.parseInt(props.getProperty("width"), -1);
@@ -618,9 +618,9 @@
                         Config.warn("Invalid connect in: " + path);
                         return false;
                     }
-                    else if (this.RenderPass > 0)
+                    else if (this.renderPass > 0)
                     {
-                        Config.warn("Render pass not supported: " + this.RenderPass);
+                        Config.warn("Render pass not supported: " + this.renderPass);
                         return false;
                     }
                     else if ((this.faces & 128) != 0)
