--- a/net/minecraft/client/renderer/chunk/RenderChunk.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunk.java
@@ -153,6 +153,7 @@
             this.chunk = null;
             this.boundingBoxParent = null;
             this.initModelviewMatrix();
+            this.needsUpdate = true;
         }
     }
 
@@ -168,7 +169,7 @@
             this.postRenderBlocks(BlockRenderLayer.TRANSLUCENT, x, y, z, bufferbuilder, compiledchunk);
         }
     }
-
+       
     public void rebuildChunk(float x, float y, float z, ChunkCompileTaskGenerator generator)
     {
         CompiledChunk compiledchunk = new CompiledChunk();
@@ -295,7 +296,7 @@
 
                 if (flag1)
                 {
-                    Reflector.callVoid(Reflector.ForgeHooksClient_setRenderLayer, null);
+                    Reflector.callVoid(Reflector.ForgeHooksClient_setRenderLayer, (BlockRenderLayer)null);
                 }
             }
 
