--- a/net/minecraft/client/renderer/chunk/RenderChunk.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunk.java
@@ -126,6 +126,7 @@
 
             this.chunk = null;
             this.initModelviewMatrix();
+            this.needsUpdate = true;
         }
     }
 
@@ -141,7 +142,7 @@
             this.postRenderBlocks(BlockRenderLayer.TRANSLUCENT, x, y, z, bufferbuilder, compiledchunk);
         }
     }
-
+       
     public void rebuildChunk(float x, float y, float z, ChunkCompileTaskGenerator generator)
     {
         CompiledChunk compiledchunk = new CompiledChunk();
@@ -177,9 +178,8 @@
             boolean flag = Reflector.ForgeBlock_canRenderInLayer.exists();
             boolean flag1 = Reflector.ForgeHooksClient_setRenderLayer.exists();
 
-            for (Object o : BlockPosM.getAllInBoxMutable(blockpos, blockpos1))
+            for (BlockPosM blockposm : (Iterable<BlockPosM>) BlockPosM.getAllInBoxMutable(blockpos, blockpos1))
             {
-            	BlockPosM blockposm = (BlockPosM) o;
                 IBlockState iblockstate = chunkcacheof.getBlockState(blockposm);
                 Block block = iblockstate.getBlock();
 
@@ -273,7 +273,7 @@
 
                 if (flag1)
                 {
-                    Reflector.callVoid(Reflector.ForgeHooksClient_setRenderLayer, null);
+                    Reflector.callVoid(Reflector.ForgeHooksClient_setRenderLayer, (BlockRenderLayer)null);
                 }
             }
 
