--- a/net/minecraft/client/entity/AbstractClientPlayer.java
+++ b/net/minecraft/client/entity/AbstractClientPlayer.java
@@ -1,8 +1,13 @@
 package net.minecraft.client.entity;
 
 import com.mojang.authlib.GameProfile;
+
 import java.io.File;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.PatreonReceiver;
+import org.vivecraft.render.PlayerModelController;
+
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.network.NetworkPlayerInfo;
 import net.minecraft.client.renderer.ImageBufferDownload;
@@ -20,6 +25,7 @@
 import net.minecraft.util.StringUtils;
 import net.minecraft.world.GameType;
 import net.minecraft.world.World;
+import net.optifine.http.FileDownloadThread;
 import net.optifine.player.CapeUtils;
 import net.optifine.player.PlayerConfigurations;
 import net.optifine.reflect.Reflector;
@@ -50,6 +56,11 @@
 
         CapeUtils.downloadCape(this);
         PlayerConfigurations.getPlayerConfiguration(this);
+        
+		    PatreonReceiver pr = new PatreonReceiver(nameClear, this);
+		    String s = "http://www.vivecraft.org/patreon/current.txt";
+		    FileDownloadThread filedownloadthread = new FileDownloadThread(s, pr);
+		    filedownloadthread.start();
     }
 
     /**
