--- a/net/minecraft/client/entity/AbstractClientPlayer.java
+++ b/net/minecraft/client/entity/AbstractClientPlayer.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.entity;
 
 import com.mojang.authlib.GameProfile;
+import org.vivecraft.api.PatreonReceiver;
+
 import java.io.File;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
@@ -50,6 +52,8 @@
 
         CapeUtils.downloadCape(this);
         PlayerConfigurations.getPlayerConfiguration(this);
+        
+        PatreonReceiver.addPlayerInfo(this);
     }
 
     /**
