--- a/net/minecraft/client/entity/AbstractClientPlayer.java
+++ b/net/minecraft/client/entity/AbstractClientPlayer.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.entity;
 
 import com.mojang.authlib.GameProfile;
+import com.mtbs3d.minecrift.api.PatreonReceiver;
+
 import java.io.File;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
@@ -16,6 +18,9 @@
 import net.minecraft.init.Items;
 import net.minecraft.src.CapeUtils;
 import net.minecraft.src.Config;
+import net.minecraft.src.FileDownloadThread;
+import net.minecraft.src.PlayerConfiguration;
+import net.minecraft.src.PlayerConfigurationReceiver;
 import net.minecraft.src.PlayerConfigurations;
 import net.minecraft.src.Reflector;
 import net.minecraft.util.ResourceLocation;
@@ -45,6 +50,11 @@
 
         CapeUtils.downloadCape(this);
         PlayerConfigurations.getPlayerConfiguration(this);
+        
+        PatreonReceiver pr = new PatreonReceiver(nameClear, this);
+        String s = "http://www.vivecraft.org/patreon/current.txt";
+        FileDownloadThread filedownloadthread = new FileDownloadThread(s, pr);
+        filedownloadthread.start();
     }
 
     /**
