--- a/net/minecraft/client/gui/GuiTextField.java
+++ b/net/minecraft/client/gui/GuiTextField.java
@@ -2,6 +2,8 @@
 
 import com.google.common.base.Predicate;
 import com.google.common.base.Predicates;
+import com.mtbs3d.minecrift.provider.MCOpenVR;
+
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.BufferBuilder;
 import net.minecraft.client.renderer.GlStateManager;
@@ -17,9 +19,10 @@
     public int x;
     public int y;
 
+    //Forge make public
     /** The width of this text field. */
-    private final int width;
-    private final int height;
+    public int width;
+    public int height;
 
     /** Has the current text being edited on the textbox. */
     private String text = "";
@@ -516,7 +519,13 @@
         if (this.canLoseFocus)
         {
             this.setFocused(flag);
-        }
+       }  else {
+	
+         }
+       	
+        // VIVE START - show keyboard
+	   if(flag && this.canLoseFocus == true) MCOpenVR.setKeyboardOverlayShowing(true, this);           // VIVE END - show keyboard
+
 
         if (this.isFocused && flag && mouseButton == 0)
         {
@@ -568,7 +577,10 @@
             if (!s.isEmpty())
             {
                 String s1 = flag ? s.substring(0, j) : s;
-                j1 = this.fontRenderer.drawStringWithShadow(s1, (float)l, (float)i1, i);
+            	j1 = this.fontRenderer.drawStringWithShadow(s1, (float)l, (float)i1, i);
+            }  else {
+            	if(!Minecraft.getMinecraft().vrSettings.seated && !MCOpenVR.keyboardShowing && this.canLoseFocus)
+            		this.fontRenderer.drawStringWithShadow("Click to open keyboard", (float)l, (float)i1, this.disabledColor);
             }
 
             boolean flag2 = this.cursorPosition < this.text.length() || this.text.length() >= this.getMaxStringLength();
@@ -732,6 +744,12 @@
         {
             Minecraft.getMinecraft().currentScreen.setFocused(isFocusedIn);
         }
+
+        // VIVE START - show keyboard
+        if(isFocusedIn && this.visible && (this.canLoseFocus == false || Minecraft.getMinecraft().vrSettings.autoOpenKeyboard))
+     	   MCOpenVR.setKeyboardOverlayShowing(isFocusedIn, this);
+        // VIVE END - show keyboard
+        
     }
 
     /**
