--- a/net/minecraft/client/gui/GuiVideoSettings.java
+++ b/net/minecraft/client/gui/GuiVideoSettings.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.gui;
 
 import java.io.IOException;
+
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.resources.I18n;
 import net.minecraft.client.settings.GameSettings;
 import net.minecraft.src.Config;
@@ -26,6 +28,10 @@
     private static final String __OBFID = "CL_00000718";
     private TooltipManager tooltipManager = new TooltipManager(this, new TooltipProviderOptions());
 
+    // Fix optifine compatiblity with hardcoredarkness
+    GuiListExtended optionsRowList = new GuiOptionsRowList(Minecraft.getMinecraft(), 0, 0, 0, 0, 0, videoOptions);
+    //
+    
     public GuiVideoSettings(GuiScreen parentScreenIn, GameSettings gameSettingsIn)
     {
         this.parentGuiScreen = parentScreenIn;
@@ -101,7 +107,7 @@
     {
         if (p_actionPerformed_1_.enabled)
         {
-            int i = this.guiGameSettings.guiScale;
+            int i = this.guiGameSettings.guiScaleUser;
 
             if (p_actionPerformed_1_.id < 200 && p_actionPerformed_1_ instanceof GuiOptionButton)
             {
@@ -115,7 +121,7 @@
                 this.mc.displayGuiScreen(this.parentGuiScreen);
             }
 
-            if (this.guiGameSettings.guiScale != i)
+            if (this.guiGameSettings.guiScaleUser != i)
             {
                 ScaledResolution scaledresolution = new ScaledResolution(this.mc);
                 int j = scaledresolution.getScaledWidth();
@@ -172,11 +178,11 @@
                     return;
                 }
 
-                if (Config.isFastRender())
-                {
-                    Config.showGuiMessage(Lang.get("of.message.shaders.fr1"), Lang.get("of.message.shaders.fr2"));
-                    return;
-                }
+//                if (Config.isFastRender())
+//                {
+//                    Config.showGuiMessage(Lang.get("of.message.shaders.fr1"), Lang.get("of.message.shaders.fr2"));
+//                    return;
+//                }
 
                 if (Config.getGameSettings().anaglyph)
                 {
@@ -243,4 +249,15 @@
     {
         return p_getGuiChatText_0_.inputField.getText();
     }
+    
+    // FORGE: fix for MC-64581 very laggy mipmap slider
+    /**
+     * Called when the screen is unloaded. Used to disable keyboard repeat events
+     */
+    @Override
+    public void onGuiClosed()
+    {
+    	super.onGuiClosed();
+    	this.mc.gameSettings.onGuiClosed();
+    }
 }
