--- a/net/minecraft/client/gui/ScaledResolution.java
+++ b/net/minecraft/client/gui/ScaledResolution.java
@@ -13,8 +13,8 @@
 
     public ScaledResolution(Minecraft minecraftClient)
     {
-        this.scaledWidth = minecraftClient.displayWidth;
-        this.scaledHeight = minecraftClient.displayHeight;
+        this.scaledWidth = 1280;
+        this.scaledHeight = 720;
         this.scaleFactor = 1;
         boolean flag = minecraftClient.isUnicode();
         int i = minecraftClient.gameSettings.guiScale;
